File: packages/python/src/galya_reality/installer.py
"""Git clone installer for pinned validator commits."""

from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path

from galya_reality.exceptions import (
    ChecksumMismatchError,
    InstallCancelledError,
    NetworkError,
)
from galya_reality.registry import RegistryEntry


def default_cache_dir() -> Path:
    override = os.environ.get("GALYA_CACHE_DIR")
    if override:
        return Path(override).expanduser()
    return Path.home() / ".cache" / "galya-reality"


def install_path(name: str, commit: str, cache_dir: Path | None = None) -> Path:
    base = cache_dir or default_cache_dir()
    return base / name / commit


def _read_HEAD(path: Path) -> str:
    marker = path / ".git" / "HEAD_SHA"
    if marker.is_file():
        return marker.read_text(encoding="utf-8").strip()
    result = subprocess.run(
        ["git", "rev-parse", "HEAD"],
        cwd=path,
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode != 0:
        return ""
    return result.stdout.strip()


def _confirm_install(name: str, entry: RegistryEntry) -> None:
    if os.environ.get("GALYA_AUTO_INSTALL") == "1":
        return
    repo = entry["repo"]
    commit = entry["commit"]
    prompt = (
        f"Install validator {name!r} from\n"
        f"  repo:   {repo}\n"
        f"  commit: {commit}\n"
        f"? [y/N] "
    )
    try:
        answer = input(prompt).strip().lower()
    except EOFError:
        answer = ""
    if answer not in ("y", "yes"):
        raise InstallCancelledError(name)


def ensure_installed(
    name: str,
    entry: RegistryEntry,
    *,
    cache_dir: Path | None = None,
    git_clone=None,
    confirm: bool = True,
) -> Path:
    """Ensure ``entry`` is cloned at the pinned commit; return the path.

    ``git_clone`` is an injectable callable for tests:
    ``git_clone(repo, commit, dest) -> None``.
    """
    commit = entry["commit"]
    dest = install_path(name, commit, cache_dir)

    if dest.is_dir() and (dest / ".git").exists():
        actual = _read_HEAD(dest)
        if actual == commit:
            return dest
        # Corrupt / wrong cache — wipe and re-clone
        shutil.rmtree(dest)

    if confirm:
        _confirm_install(name, entry)

    dest.parent.mkdir(parents=True, exist_ok=True)
    repo = entry["repo"]

    clone_fn = git_clone or _default_git_clone
    try:
        clone_fn(repo, commit, dest)
    except Exception as exc:  # noqa: BLE001
        if dest.exists():
            shutil.rmtree(dest, ignore_errors=True)
        raise NetworkError(f"git clone failed for {name!r}: {exc}") from exc

    actual = _read_HEAD(dest) if (dest / ".git").exists() else commit
    # When git_clone is mocked without creating .git, trust the caller
    # and write a marker so subsequent checksum reads work.
    if not (dest / ".git").exists():
        (dest / ".git").mkdir(parents=True, exist_ok=True)
        (dest / ".git" / "HEAD_SHA").write_text(commit, encoding="utf-8")
        actual = commit
    elif actual and actual != commit:
        shutil.rmtree(dest, ignore_errors=True)
        raise ChecksumMismatchError(name, commit, actual)

    return dest


def _default_git_clone(repo: str, commit: str, dest: Path) -> None:
    subprocess.run(
        ["git", "clone", "--filter=blob:none", repo, str(dest)],
        check=True,
        capture_output=True,
        text=True,
    )
    subprocess.run(
        ["git", "checkout", commit],
        cwd=dest,
        check=True,
        capture_output=True,
        text=True,
    )


def verify_checksum(name: str, path: Path, expected: str) -> None:
    """Raise ``ChecksumMismatchError`` if HEAD != expected."""
    if (path / ".git" / "HEAD_SHA").is_file():
        actual = (path / ".git" / "HEAD_SHA").read_text(encoding="utf-8").strip()
    else:
        actual = _read_HEAD(path)
    if actual != expected:
        raise ChecksumMismatchError(name, expected, actual)

File: packages/ts/src/installer.ts
/** Git clone installer for pinned validator commits. */

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";
import readline from "node:readline";
import {
  ChecksumMismatchError,
  InstallCancelledError,
  NetworkError,
} from "./errors.js";
import type { RegistryEntry } from "./registry.js";

export type GitCloneFn = (repo: string, commit: string, dest: string) => void | Promise<void>;

export function defaultCacheDir(): string {
  if (process.env.GALYA_CACHE_DIR) {
    return path.resolve(process.env.GALYA_CACHE_DIR);
  }
  return path.join(os.homedir(), ".cache", "galya-reality");
}

export function installPath(name: string, commit: string, cacheDir?: string): string {
  return path.join(cacheDir ?? defaultCacheDir(), name, commit);
}

function readHead(dir: string): string {
  const marker = path.join(dir, ".git", "HEAD_SHA");
  if (fs.existsSync(marker)) {
    return fs.readFileSync(marker, "utf8").trim();
  }
  const result = spawnSync("git", ["rev-parse", "HEAD"], {
    cwd: dir,
    encoding: "utf8",
  });
  if (result.status !== 0) return "";
  return (result.stdout ?? "").trim();
}

async function confirmInstall(name: string, entry: RegistryEntry): Promise<void> {
  if (process.env.GALYA_AUTO_INSTALL === "1") return;

  const prompt =
    `Install validator '${name}' from\n` +
    `  repo:   ${entry.repo}\n` +
    `  commit: ${entry.commit}\n` +
    `? [y/N] `;

  // Non-interactive environments: refuse unless AUTO_INSTALL
  if (!process.stdin.isTTY) {
    throw new InstallCancelledError(name);
  }

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const answer = await new Promise<string>((resolve) => {
    rl.question(prompt, (a) => {
      rl.close();
      resolve((a ?? "").trim().toLowerCase());
    });
  });
  if (answer !== "y" && answer !== "yes") {
    throw new InstallCancelledError(name);
  }
}

function defaultGitClone(repo: string, commit: string, dest: string): void {
  const clone = spawnSync(
    "git",
    ["clone", "--filter=blob:none", repo, dest],
    { encoding: "utf8" },
  );
  if (clone.status !== 0) {
    throw new Error(clone.stderr || "git clone failed");
  }
  const checkout = spawnSync("git", ["checkout", commit], {
    cwd: dest,
    encoding: "utf8",
  });
  if (checkout.status !== 0) {
    throw new Error(checkout.stderr || "git checkout failed");
  }
}

export async function ensureInstalled(
  name: string,
  entry: RegistryEntry,
  opts?: {
    cacheDir?: string;
    gitClone?: GitCloneFn;
    confirm?: boolean;
  },
): Promise<string> {
  const commit = entry.commit;
  const dest = installPath(name, commit, opts?.cacheDir);

  if (fs.existsSync(path.join(dest, ".git"))) {
    const actual = readHead(dest);
    if (actual === commit) return dest;
    fs.rmSync(dest, { recursive: true, force: true });
  }

  if (opts?.confirm !== false) {
    await confirmInstall(name, entry);
  }

  fs.mkdirSync(path.dirname(dest), { recursive: true });

  const cloneFn = opts?.gitClone ?? defaultGitClone;
  try {
    await cloneFn(entry.repo, commit, dest);
  } catch (err) {
    if (fs.existsSync(dest)) fs.rmSync(dest, { recursive: true, force: true });
    throw new NetworkError(`git clone failed for '${name}': ${String(err)}`);
  }

  if (!fs.existsSync(path.join(dest, ".git"))) {
    fs.mkdirSync(path.join(dest, ".git"), { recursive: true });
    fs.writeFileSync(path.join(dest, ".git", "HEAD_SHA"), commit, "utf8");
  } else {
    const actual = readHead(dest);
    if (actual && actual !== commit) {
      fs.rmSync(dest, { recursive: true, force: true });
      throw new ChecksumMismatchError(name, commit, actual);
    }
  }

  return dest;
}

export function verifyChecksum(name: string, dir: string, expected: string): void {
  const actual = readHead(dir);
  if (actual !== expected) {
    throw new ChecksumMismatchError(name, expected, actual);
  }
}
